### PR TITLE
filtering mapped port and ip for proxy

### PR DESCRIPTION
**Pull request by allopez7**

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -77,8 +77,16 @@ if (process.env.NODE_ENV === 'production') {
...
-      res.end('hello world')
+    const [totalDomain, mappings] = [(req.headers.host).split('.'), getMappings()]
+    const [domain, subdomain] = [totalDomain.slice(-2).join(''), totalDomain[0]]
+    let mappedDomain;
+    if(totalDomain.length === 3){
+    	mappedDomain = mappings.filter(e=>e.subDomain === subdomain && e.domain === domain)
     }
+	mappedDomain = domain
+	const {ip, port} = mappedDomain
+	console.log('ip', ip, 'port', port)
```
Comment (allopez7) [inline src/app.ts]: removing

File: src/app.ts [hunk 2]
```diff
@@ -77,8 +77,16 @@ if (process.env.NODE_ENV === 'production') {
       }
     },
     (req, res) => {
-      res.end('hello world')
+    const [totalDomain, mappings] = [(req.headers.host).split('.'), getMappings()]
+    const [domain, subdomain] = [totalDomain.slice(-2).join(''), totalDomain[0]]
+    let mappedDomain;
+    if(totalDomain.length === 3){
+    	mappedDomain = mappings.filter(e=>e.subDomain === subdomain && e.domain === domain)
```
Comment (allopez7) [inline src/app.ts]: will add "find"

File: src/app.ts [hunk 3]
```diff
@@ -77,8 +77,16 @@ if (process.env.NODE_ENV === 'production') {
...
     }
+	mappedDomain = domain
```
Comment (allopez7) [inline src/app.ts]: domain is a string but should be an object. Will fix this issue.

File: src/app.ts [hunk 4]
```diff
@@ -77,6 +80,24 @@ if (process.env.NODE_ENV === 'production') {
...
+      ]
+      let mappedDomain
+      if (totalDomain.length === 3) {
+        mappedDomain = mappings.find(
+          e => e.subDomain === subdomain && e.domain === domain
+        )
+      }
+      if(totalDomain.length === 2){
+      	mappedDomain = mappings.find(e=>e.domain === domain && !e.subDomain)
+      }
```
Comment (songz) [inline src/app.ts]: All these lines of code can be condensed into 3 lines: 
```
const {ip, port} = mappings.find( ({subDomain, domain}) => {
  return `${subDomain}.${domain}` === req.headers.host
}) || {}
```

File: src/app.ts [hunk 5]
```diff
@@ -87,7 +90,16 @@ if (process.env.NODE_ENV === 'production') {
...
     },
     (req, res) => {
-      res.end('hello world')
+      try {
+        const mappings = getMappings()
+        const { ip, port } =
+          mappings.find(({ subDomain, domain }) => {
+            return `${subDomain}.${domain}` === req.headers.host
+          }) || {}
+        if (port) return proxy.web(req, res, { target: `${ip}:${port}` })
```
Comment (songz) [inline src/app.ts]: Please handle errors gracefully.
```
if (!port) return res.send('Cannot redirect')
return proxy.web....
```

Comment (songz) [reply src/app.ts]: and then please add error handler. When there is an error, please do `res.send('Error communicating with server that runs <hostname>')`

File: src/app.ts [hunk 6]
```diff
@@ -87,10 +92,31 @@ if (process.env.NODE_ENV === 'production') {
...
     (req, res) => {
-      res.end('hello world')
+      try {
+        const mappings = getMappings()
+        const { ip, port }: ProxyMapping =
+          mappings.find(({ subDomain, domain }) => {
+            return `${subDomain}.${domain}` === req.headers.host
+          }) || {}
+        if (!port) return res.end('Cannot redirect')
+        if (port)
```
Comment (songz) [inline src/app.ts]: remove. unnecessary, because above you returned.

File: src/app.ts [hunk 7]
```diff
@@ -87,10 +92,31 @@ if (process.env.NODE_ENV === 'production') {
...
     },
     (req, res) => {
-      res.end('hello world')
+      try {
+        const mappings = getMappings()
+        const { ip, port }: ProxyMapping =
+          mappings.find(({ subDomain, domain }) => {
+            return `${subDomain}.${domain}` === req.headers.host
+          }) || {}
+        if (!port) return res.end('Cannot redirect')
```
Comment (songz) [inline src/app.ts]: You are not redirecting. `res.end('Not Found')`

Comment (songz) [reply src/app.ts]: Also, `if (!port || !ip)`

File: src/app.ts [hunk 8]
```diff
@@ -8,8 +8,15 @@ import { hashPass } from './helpers/crypto'
...
 import tls from 'tls'
-import { getAvailableDomains } from './lib/data'
+import { getAvailableDomains, getMappings } from './lib/data'
 import { isCorrectCredentials } from './auth'
+import httpProxy from 'http-proxy'
+import { ProxyMapping } from './types/general'
+
+const proxy = httpProxy.createProxyServer({})
+proxy.on('error', err => {
+  console.log('error', err)
```
Comment (songz) [inline src/app.ts]: could you put in a better log message? 
```
console.error('proxy error', err)
```

File: src/app.ts [hunk 9]
```diff
@@ -87,10 +92,25 @@ if (process.env.NODE_ENV === 'production') {
...
-      res.end('hello world')
+      try {
+        const mappings = getMappings()
+        const { ip, port }: ProxyMapping =
+          mappings.find(({ subDomain, domain }) => {
+            return `${subDomain}.${domain}` === req.headers.host
+          }) || {}
+        if (!port || !ip) return res.end('Not Found')
+        proxy.web(req, res, { target: `http://${ip}:${port}` }, err => {
+          console.log('Error communicating with server', err)
```
Comment (songz) [inline src/app.ts]: `console.error`

File: src/app.ts [hunk 10]
```diff
@@ -87,10 +92,25 @@ if (process.env.NODE_ENV === 'production') {
...
+          res.end(
+            `Error communicating with server that runs ${req.headers.host}`
+          )
+        })
+      } catch (err) {
+        console.log('Error: proxy failed', err)
```
Comment (songz) [inline src/app.ts]: `console.error`

File: src/app.ts [hunk 11]
```diff
@@ -8,8 +8,15 @@ import { hashPass } from './helpers/crypto'
...
 import tls from 'tls'
-import { getAvailableDomains } from './lib/data'
+import { getAvailableDomains, getMappings } from './lib/data'
 import { isCorrectCredentials } from './auth'
+import httpProxy from 'http-proxy'
+import { ProxyMapping } from './types/general'
+
+const proxy = httpProxy.createProxyServer({})
+proxy.on('error', err => {
+  console.log('Proxy error', err)
```
Comment (songz) [inline src/app.ts]: `console.error`

**Pull request merged**